File: java project/project/models/Complaint.java
package project.models;

/**
 * Represents a complaint submitted by a citizen.
 * This class stores complaint details and supports file-based storage.
 */
public class Complaint {

    private String complaintId;
    private String username;
    private String topic;
    private String priority;
    private String description;
    private String location;
    private String department;
    private String date;
    private String time;
    private String status;
    private String adminRemark;

    /**
     * Creates an empty complaint.
     */
    public Complaint() {
    }

    /**
     * Creates a complaint with all details.
     *
     * @param complaintId unique complaint ID
     * @param username owner username
     * @param topic complaint topic
     * @param priority priority level
     * @param description complaint description
     * @param location complaint location
     * @param department concerned department
     * @param date complaint date
     * @param time complaint time
     * @param status current status
     * @param adminRemark remark from admin
     */
    public Complaint(String complaintId, String username, String topic, String priority, String description,
                     String location, String department, String date, String time, String status, String adminRemark) {
        this.complaintId = complaintId;
        this.username = username;
        this.topic = topic;
        this.priority = priority;
        this.description = description;
        this.location = location;
        this.department = department;
        this.date = date;
        this.time = time;
        this.status = status;
        this.adminRemark = adminRemark;
    }

    public String getComplaintId() {
        return complaintId;
    }

    public void setComplaintId(String complaintId) {
        this.complaintId = complaintId;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getTopic() {
        return topic;
    }

    public void setTopic(String topic) {
        this.topic = topic;
    }

    public String getPriority() {
        return priority;
    }

    public void setPriority(String priority) {
        this.priority = priority;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public String getDepartment() {
        return department;
    }

    public void setDepartment(String department) {
        this.department = department;
    }

    public String getDate() {
        return date;
    }

    public void setDate(String date) {
        this.date = date;
    }

    public String getTime() {
        return time;
    }

    public void setTime(String time) {
        this.time = time;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String getAdminRemark() {
        return adminRemark;
    }

    public void setAdminRemark(String adminRemark) {
        this.adminRemark = adminRemark;
    }

    /**
     * Converts the complaint to a pipe-separated line for complaints.txt.
     *
     * @return complaint line for file storage
     */
    public String toFileString() {
        return complaintId + "|" + username + "|" + topic + "|" + priority + "|" + description + "|"
                + location + "|" + department + "|" + date + "|" + time + "|" + status + "|" + adminRemark;
    }

    /**
     * Builds a Complaint object from one file line in complaints.txt.
     *
     * @param line stored complaint line
     * @return Complaint object, or null if the line is invalid
     */
    public static Complaint fromFileString(String line) {
        if (line == null || line.trim().isEmpty()) {
            return null;
        }

        String[] parts = line.split("\\|", -1);
        if (parts.length < 11) {
            return null;
        }

        return new Complaint(
                parts[0].trim(),
                parts[1].trim(),
                parts[2].trim(),
                parts[3].trim(),
                parts[4].trim(),
                parts[5].trim(),
                parts[6].trim(),
                parts[7].trim(),
                parts[8].trim(),
                parts[9].trim(),
                parts[10].trim()
        );
    }
}

File: cpp_project/models/Complaint.cpp
#include "Complaint.h"
#include <sstream>
#include <vector>

namespace project {
namespace models {

Complaint::Complaint() : complaintId("") {}

Complaint::Complaint(std::string complaintId, std::string username, std::string topic, std::string priority,
                     std::string description, std::string location, std::string department,
                     std::string date, std::string time, std::string status, std::string adminRemark)
    : complaintId(complaintId), username(username), topic(topic), priority(priority),
      description(description), location(location), department(department), date(date),
      time(time), status(status), adminRemark(adminRemark) {}

std::string Complaint::getComplaintId() const { return complaintId; }
void Complaint::setComplaintId(const std::string& id) { complaintId = id; }

std::string Complaint::getUsername() const { return username; }
void Complaint::setUsername(const std::string& username) { this->username = username; }

std::string Complaint::getTopic() const { return topic; }
void Complaint::setTopic(const std::string& topic) { this->topic = topic; }

std::string Complaint::getPriority() const { return priority; }
void Complaint::setPriority(const std::string& priority) { this->priority = priority; }

std::string Complaint::getDescription() const { return description; }
void Complaint::setDescription(const std::string& description) { this->description = description; }

std::string Complaint::getLocation() const { return location; }
void Complaint::setLocation(const std::string& location) { this->location = location; }

std::string Complaint::getDepartment() const { return department; }
void Complaint::setDepartment(const std::string& department) { this->department = department; }

std::string Complaint::getDate() const { return date; }
void Complaint::setDate(const std::string& date) { this->date = date; }

std::string Complaint::getTime() const { return time; }
void Complaint::setTime(const std::string& time) { this->time = time; }

std::string Complaint::getStatus() const { return status; }
void Complaint::setStatus(const std::string& status) { this->status = status; }

std::string Complaint::getAdminRemark() const { return adminRemark; }
void Complaint::setAdminRemark(const std::string& adminRemark) { this->adminRemark = adminRemark; }

std::string Complaint::toFileString() const {
    return complaintId + "|" + username + "|" + topic + "|" + priority + "|" + description + "|" +
           location + "|" + department + "|" + date + "|" + time + "|" + status + "|" + adminRemark;
}

Complaint Complaint::fromFileString(const std::string& line) {
    if (line.empty()) return Complaint();

    std::vector<std::string> parts;
    std::stringstream ss(line);
    std::string part;

    while (std::getline(ss, part, '|')) {
        parts.push_back(part);
    }
    // Handle trailing empty parts
    if (!line.empty() && line.back() == '|') {
        parts.push_back("");
    }

    if (parts.size() < 11) return Complaint();

    return Complaint(parts[0], parts[1], parts[2], parts[3], parts[4],
                     parts[5], parts[6], parts[7], parts[8], parts[9], parts[10]);
}

bool Complaint::isValid() const {
    return !complaintId.empty();
}

}
}
